### ISSUE TITLE
Just a question

### ISSUE THREAD

**Issue by evermeer** [ishkawa/DataSourceKit#4]
Maybe I'm missing something but... 

The dataSource declaration seems to be a straight forward mapping from enum value to cell type which always have the same structure. I think it's possible to move that mapping to the CellsDeclarator (and CellDeclaration) like the code shown below. You just have to make sure the makeBinder function is called when you call the cell function. This would make the code cleaner and even more compact.

```
    enum CellDeclaration: Equatable {
        case outline(BindableCell.Type, Venue)
        case sectionHeader(BindableCell.Type, String)
        case review(BindableCell.Type, Review)
        case relatedVenue(BindableCell.Type, Venue)
    }
    
    struct Data: CellsDeclarator {
        var venue: Venue
        var reviews: [Review]
        var relatedVenues: [Venue]

        func declareCells(_ cell: (CellDeclaration) -> Void) {
            cell(.outline(VenueOutlineCell.self, venue))

            if !reviews.isEmpty {
                cell(.sectionHeader(SectionHeaderCell.self, "Reviews"))
                for review in reviews {
                    cell(.review(ReviewCell.self, review))
                }
            }
            
            if !relatedVenues.isEmpty {
                cell(.sectionHeader(SectionHeaderCell.self, "Related Venues"))
                for relatedVenue in relatedVenues {
                    cell(.relatedVenue(RelatedVenueCell.self, relatedVenue))
                }
            }
        }
    }
```

**Comment by ishkawa** [comment]
Yes! Code you suggested is apparently more compact, but it does 2 things at a time.

1. Transform `Data` into form suitable for collection.
2. Associates the elements with UICollectionViewCell subclasses.

IMO, these steps should be separated so that artifact of the first step is isolated from UIKit. As a result of the separation, we can test that `CellsDeclarator` creates `[CellDeclaration]` correctly like below. 

```swift
// https://github.com/ishkawa/DataSourceKit/blob/b2d1dcc71c3ae1e4599291690c7e15f2b0733909/DemoTests/VenuDetailViewControllerTests.swift#L13-L32
func testEmptyRelatedVenues() {
    let venue = Venue(photo: nil, name: "Kaminarimon")
    let review1 = Review(authorImage: nil, authorName: "Yosuke Ishikawa", body: "Foo")
    let review2 = Review(authorImage: nil, authorName: "Masatake Yamoto", body: "Bar")
    
    let data = VenueDetailViewController.Data(
        venue: venue,
        reviews: [
            review1,
            review2,
        ],
        relatedVenues: [])
    
    XCTAssertEqual(data.cellDeclarations, [
        .outline(venue),
        .sectionHeader("Reviews"),
        .review(review1),
        .review(review2),
    ])
}
```

It's easy to write tests because `[CellDeclaration]` is just an alternative form of `Data`. If they have also things from UIKit, the test gets more complicated. I think ease of testing is more important than ease of writing view code.

The separation may seem to be weird if the 2 steps are written in the same place (like `VenueDetailViewController` in demo app). On the other hand, if view controller is separated into state and layout like MVVM, the separation is straightforward.

**Comment by evermeer** [comment]
Ah, true, you are right. At the other hand I do love compact code and I hate repetitions. Maybe support both? 🙂If the first item of the associated value is a BindableCell.Type then... I think you can get the associated value directly in Swift 4.2. If you do want to support older version, then you could use: https://github.com/evermeer/Stuff#enum

**Comment by ishkawa** [comment]
After some consideration, I found that it's possible to support both with following extension. 

```swift
extension CollectionViewDataSource where CellDeclaration == CellBinder {
    // Add meaningless parameter to avoid error `Members of constrained extensions
    // cannot be declared @objc`, which is caused by collision to init() of NSObject.
    public convenience init(binderFromBinder: @escaping (CellBinder) -> CellBinder = { $0 }) {
        self.init(binderFromDeclaration: binderFromBinder)
    }
}
```

When we use `CellBinder` for `CellDeclaration`, `enum CellDeclaration` and `binderFromDeclaration`  can be omitted. In this case, The only thing to create data source is `declareCells(_:)` like below.

```swift
func declareCells(_ cell: (CellBinder) -> Void) {
    cell(VenueOutlineCell.makeBinder(value: venue))
    
    if !reviews.isEmpty {
        cell(SectionHeaderCell.makeBinder(value: "Reviews"))
        for review in reviews {
            cell(ReviewCell.makeBinder(value: review))
        }
    }
    
    if !relatedVenues.isEmpty {
        cell(SectionHeaderCell.makeBinder(value: "Related Venues"))
        for relatedVenue in relatedVenues {
            cell(RelatedVenueCell.makeBinder(value: relatedVenue))
        }
    }
}
```

[Here](https://github.com/ishkawa/DataSourceKit/commit/eead977513c938e601b0130ec634aaa14f4b714a)'s full difference between current demo and the new demo.

Existing style is still available, but I came to think that the new version is better for more people who don't need the separation. So I revised README in #7 to encourage people to use the new style for the first time.

**Comment by evermeer** [comment]
Extremely awesome! I will play with it later today!

**Comment by ishkawa** [comment]
Just merged #7 and released it in 0.2.0!
